### ISSUE TITLE
`torch.compile` produces wrong results for `fractional_max_pool2d` when `_random_samples` is not provided

### ISSUE THREAD

**Issue by wuyii8941** [pytorch/pytorch#183249]
## 🐛 Describe the bug

`torch.compile(backend='inductor')` produces incorrect output for `torch.nn.functional.fractional_max_pool2d` when the internal random sampling path is used (i.e., `_random_samples` is not explicitly passed). The compiled version selects **different pooling indices** than eager mode, returning wrong values.

This is **not** a numerical precision issue — the compiled kernel picks entirely different source elements. With `_random_samples` explicitly provided, the results match exactly.

## Reproducer

```python
import torch
import torch._inductor
torch._inductor.config.force_disable_caches = True

x = torch.randn(1, 3, 16, 16, device='cuda')

# Eager
torch.manual_seed(42); torch.cuda.manual_seed(42)
eager_out, eager_idx = torch.nn.functional.fractional_max_pool2d(
    x, 3, output_size=(4, 4), return_indices=True)

# Compiled
torch._dynamo.reset()
torch.manual_seed(42); torch.cuda.manual_seed(42)
comp_out, comp_idx = torch.compile(
    lambda x: torch.nn.functional.fractional_max_pool2d(
        x, 3, output_size=(4, 4), return_indices=True),
    backend='inductor'
)(x)

print(f"Index mismatches: {(eager_idx != comp_idx).sum().item()}/{eager_idx.numel()}")
print(f"Value max_diff: {(eager_out - comp_out).abs().max().item():.4f}")
# Index mismatches: 6/48
# Value max_diff: 1.2810
```

**Workaround**: passing `_random_samples` explicitly produces correct results:

```python
samples = torch.rand(1, 3, 2, device='cuda')
# Both eager and compiled produce identical output when _random_samples is given
eager = torch.nn.functional.fractional_max_pool2d(x, 3, output_size=(4, 4), _random_samples=samples)[0]
compiled = torch.compile(
    lambda x: torch.nn.functional.fractional_max_pool2d(
        x, 3, output_size=(4, 4), _random_samples=samples)[0],
    backend='inductor'
)(x)
print((eager - compiled).abs().max().item())  # 0.0
```

## Observations

| Condition | Result |
|-----------|--------|
| Without `_random_samples` (default) | **Wrong**: 6/48 indices differ, `max_diff=1.28` |
| With explicit `_random_samples` | Correct: `max_diff=0.0` |
| Eager same-seed deterministic | Yes |
| Compiled same-seed deterministic | Yes |
| Eager == Compiled (same seed) | **No** — different indices selected |
| Multiple calls to compiled fn | Each call uses random state (not baked-in) |

The compiled function does consume random state across calls (different seeds produce different outputs), but the random state is consumed differently than in eager mode, causing the pooling regions to diverge.

## Expected behavior

`torch.compile` should produce the same output as eager mode for the same random state. `fractional_max_pool2d` without `_random_samples` should be functionally equivalent between eager and compiled execution.

## Environment

- PyTorch: 2.13.0.dev20260501+cu126
- GPU: Tesla T4
- CUDA: 12.6
- OS: Linux

cc @pbelevich @mikaylagawarecki @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by karthickai** [comment]
**🤖 PT2 Repro Agent**

I tried to reproduce this issue on PyTorch nightly (`2.13.0.dev20260518+cu130`) on NVIDIA H100 and confirmed it reproduces (3/3 runs failed).

All 3 runs reproduce the reported behavior: 6/48 index mismatches between eager and inductor fractional_max_pool2d (output_size=(4,4)), with value max_diff=1.351943. allclose fails at every tested tolerance from 1e-4 down to 1e-8. Passing _random_samples explicitly produces max_diff=0.0, matching the documented workaround. Re-running with TORCHINDUCTOR_EMULATE_PRECISION_CASTS=1 did not change the outcome.

The compiler bisector identified the offending component:
- **Backend**: `inductor`
- **Subsystem**: `inductor_fallback_random`
- **Detail**: `inductor_fallback_random`

<details>
<summary>Per-subsystem coverage (4 tested)</summary>

| Subsystem | Bug fixed when disabled? |
|---|---|
| `pre_grad_passes` | ❌ no |
| `joint_graph_passes` | ❌ no |
| `post_grad_passes` | ❌ no |
| `inductor_fallback_random` | ✅ yes |

Replay: `TORCH_BISECT_BACKEND=inductor TORCH_BISECT_SUBSYSTEM=inductor_fallback_random TORCH_BISECT_MAX=0 python repro.py`

_Tool: `torch._inductor.compiler_bisector`_
</details>

<details>
<summary>Error output</summary>

```
PyTorch: 2.13.0.dev20260518+cu130
CUDA available: True
GPU: NVIDIA H100
Index mismatches: 6/48
Value max_diff:   1.351943
  allclose(atol=0.0001): False
  allclose(atol=1e-05): False
  allclose(atol=1e-06): False
  allclose(atol=1e-07): False
  allclose(atol=1e-08): False
Workaround (_random_samples) max_diff: 0.000000

/home/karthickps/.conda/envs/pytorch-nightly/lib/python3.13/site-packages/torch/_dynamo/pgo.py:596: UserWarning: dynamo_pgo force disabled by torch.compiler.config.force_disable_caches
  warn_once(
/home/karthickps/.conda/envs/pytorch-nightly/lib/python3.13/site-packages/torch/_dynamo/pgo.py:596: UserWarning: dynamo_pgo force disabled by torch.compiler.config.force_disable_caches
  warn_once(
/home/karthickps/.conda/envs/pytorch-nightly/lib/python3.13/site-packages/torch/_dynamo/pgo.py:596: UserWarning: dynamo_pgo force disabled by torch.compiler.config.force_disable_caches
  warn_once(

ERROR conda.cli.main_run:execute(49): `conda run python repro.py` failed. (See above for error)

```
</details>

<details>
<summary>Repro script I used</summary>

```python
import os
import torch
import torch._inductor

torch._inductor.config.force_disable_caches = True


def _run(backend: str):
    device = "cuda" if torch.cuda.is_available() else "cpu"

    torch.manual_seed(0)
    if device == "cuda":
        torch.cuda.manual_seed(0)
    x = torch.randn(1, 3, 16, 16, device=device)

    # Eager
    torch.manual_seed(42)
    if device == "cuda":
        torch.cuda.manual_seed(42)
    eager_out, eager_idx = torch.nn.functional.fractional_max_pool2d(
        x, 3, output_size=(4, 4), return_indices=True
    )

    # Compiled
    torch._dynamo.reset()
    torch.manual_seed(42)
    if device == "cuda":
        torch.cuda.manual_seed(42)
    comp_out, comp_idx = torch.compile(
        lambda inp: torch.nn.functional.fractional_max_pool2d(
            inp, 3, output_size=(4, 4), return_indices=True
        ),
        backend=backend,
    )(x)

    return x, eager_out, eager_idx, comp_out, comp_idx


def test_fn() -> bool:
    """Return True if NO bug (eager == compiled). False if bug present."""
    torch._dynamo.reset()
    backend = os.environ.get("TORCH_COMPILE_BACKEND", "inductor")
    _, eager_out, eager_idx, comp_out, comp_idx = _run(backend)
    idx_match = bool((eager_idx == comp_idx).all().item())
    value_match = torch.allclose(eager_out, comp_out, atol=1e-5)
    return idx_match and value_match


if __name__ == "__main__":
    print(f"PyTorch: {torch.__version__}")
    print(f"CUDA available: {torch.cuda.is_available()}")
    if torch.cuda.is_available():
        print(f"GPU: {torch.cuda.get_device_name(0)}")

    try:
        x, eager_out, eager_idx, comp_out, comp_idx = _run("inductor")
        idx_mismatches = (eager_idx != comp_idx).sum().item()
        max_diff = (eager_out - comp_out).abs().max().item()
        print(f"Index mismatches: {idx_mismatches}/{eager_idx.numel()}")
        print(f"Value max_diff:   {max_diff:.6f}")

        for atol in (1e-4, 1e-5, 1e-6, 1e-7, 1e-8):
            print(
                f"  allclose(atol={atol:g}): "
                f"{torch.allclose(eager_out, comp_out, atol=atol)}"
            )

        device = "cuda" if torch.cuda.is_available() else "cpu"
        samples = torch.rand(1, 3, 2, device=device)
        eager_ws = torch.nn.functional.fractional_max_pool2d(
            x, 3, output_size=(4, 4), _random_samples=samples
        )[0]
        torch._dynamo.reset()
        comp_ws = torch.compile(
            lambda inp: torch.nn.functional.fractional_max_pool2d(
                inp, 3, output_size=(4, 4), _random_samples=samples
            )[0],
            backend="inductor",
        )(x)
        print(
            f"Workaround (_random_samples) max_diff: "
            f"{(eager_ws - comp_ws).abs().max().item():.6f}"
        )

        ok = test_fn()
    except Exception:
        import traceback

        traceback.print_exc()
        os._exit(2)
    os._exit(0 if ok else 1)

```
</details>

<details>
<summary>My environment</summary>

```
Collecting environment information...
PyTorch version: 2.13.0.dev20260518+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: CentOS Stream 9 (x86_64)
GCC version: (GCC) 11.5.0 20240719 (Red Hat 11.5.0-14)
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.34

Python version: 3.13.12 | packaged by Anaconda, Inc. | (main, Feb 24 2026, 16:13:31) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.16.1-0_fbk2_hardened_0_gf40efc324cc8-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA H100
Nvidia driver version: 580.126.09
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  46
On-line CPU(s) list:                     0-45
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 9654 96-Core Processor
CPU family:                              25
Model:                                   17
Thread(s) per core:                      1
Core(s) per socket:                      46
Socket(s):                               1
Stepping:                                1
BogoMIPS:                                4792.79
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm rep_good nopl cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw perfctr_core ssbd ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512_bf16 clzero xsaveerptr wbnoinvd arat npt lbrv nrip_save tsc_scale vmcb_clean flushbyasid pausefilter pfthreshold v_vmsave_vmload vgif vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid fsrm flush_l1d arch_capabilities
Virtualization:                          AMD-V
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               2.9 MiB (46 instances)
L1i cache:                               2.9 MiB (46 instances)
L2 cache:                                23 MiB (46 instances)
L3 cache:                                736 MiB (46 instances)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-45
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Old microcode:             Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Vulnerable
Vulnerability Spec store bypass:         Vulnerable
Vulnerability Spectre v1:                Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:                Vulnerable; IBPB: disabled; STIBP: disabled; PBRSB-eIBRS: Not affected; BHI: Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable
Vulnerability Tsx async abort:           Not affected

Versions of relevant libraries:
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.4.4
[pip3] nvidia-cublas==13.1.1.3
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvtx==13.0.85
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] onnx==1.21.0
[pip3] onnx-ir==0.2.1
[pip3] onnxscript==0.7.0
[pip3] torch==2.13.0.dev20260517+cu130
[pip3] torchao==0.17.0
[pip3] torchaudio==2.11.0.dev20260517+cu130
[pip3] torchdata==0.11.0
[pip3] torchinfo==1.8.0
[pip3] torchmetrics==1.9.0
[pip3] torchrec==1.7.0.dev20260413+cu130
[pip3] torchtune==0.6.1
[pip3] torchvision==0.28.0.dev20260518+cu130
[pip3] triton==3.7.0+git88b227e2
[conda] Could not collect
```
</details>

---
If the repro agent produced an incorrect result (wrong classification, bad repro script, missed the actual bug, etc.), add the label `repro:agent:h100:incorrect` to the issue.
